tests/unit/compiler/venom/test_common_subexpression_elimination.py

```diff
@@ -0,0 +1,535 @@
+    _check_pre_post(pre, pre)
+
+
+def _check_no_change_fn(pre: str):
+    _check_pre_post_fn(pre, pre)
+
+
+def test_common_subexpression_elimination():
+    pre = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = add %1, 10
+        %mul2 = mul %sum2, 10
+        return %mul1, %mul2
+    """
+
+    post = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = %sum1
+        %mul2 = %mul1
+        return %mul1, %mul2
+    """
+
+    _check_pre_post(pre, post)
+
+
+def test_common_subexpression_elimination_commutative():
+    pre = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = add 10, %1
+        %mul2 = mul 10, %sum2
+        return %mul1, %mul2
+    """
+
+    post = """
+    main:
+        %1 = param
+        %sum1 = add %1, 10
+        %mul1 = mul %sum1, 10
+        %sum2 = %sum1
+        %mul2 = %mul1
+        return %mul1, %mul2
+    """
+
+    _check_pre_post(pre, post)
+
+
+def test_common_subexpression_elimination_no_commutative():
+    """
+    Tests you cannot substitute cummutated non-cumutative operations
+    """
+    pre = """
+    main:
+        %1 = param
+        %sum1 = sub %1, 10
+        %sum2 = sub 10, %1
+        return %sum1, %sum2
+    """
+
+    _check_no_change(pre)
+
+
+def test_common_subexpression_elimination_effects_1():
+    """
+    Test that inner dependencies have correct barrier
+    """
+    pre = """
+    main:
+        %par = param
+        %mload1 = mload 0
+
+        ; barrier
+        mstore 0, %par
+        %mload2 = mload 0
+
+        ; adds cannot be substituted because the
+        ; mloads are separated with barrier
+        %1 = add %mload1, 10
+        %2 = add %mload2, 10
+        return %1, %2
+    """
+
+    _check_no_change(pre)
+
+
+def test_common_subexpression_elimination_effects_2():
+    """
+    Test that barrier does not affect inner dependencies
+    """
+    pre = """
+    main:
+        %par = param
+        %mload1 = mload 0
+        %1 = add %mload1, 10
+
+        ; barrier
+        mstore 0, %par
+        %mload2 = mload 0
+
+        ; mload1 is still valid and same
+        %2 = add %mload1, 10
+
+        ; this cannot be substituted since mload1
+        ; and mload2 are separated by barrier
+        %3 = add %mload2, 10
+        return %1, %2
+    """
+
+    post = """
+    main:
+        %par = param
+        %mload1 = mload 0
+        %1 = add %mload1, 10
+        mstore 0, %par
+        %mload2 = mload 0
+        %2 = %1
+        %3 = add %mload2, 10
+        return %1, %2
+    """
+
+    _check_pre_post(pre, post)
+
+
+def test_common_subexpression_elimination_logs_no_indepontent():
+    pre = """
+    main:
+        %1 = 10
+        log %1
+        log %1
+        stop
+    """
+
+    _check_no_change(pre)
+
+
+def test_common_subexpression_elimination_effects_3():
```

Comment:
slightly shorter names will be helpful when selecting tests manually, e.g. `pytest -k test_cse_elim_effects3`